util/svdgen.py

```diff
@@ -0,0 +1,604 @@
+    """Generate an SVD <device> node from the top-level HJSON. For each
+    module defined in `top` the corresponding peripheral definition is
+    generated by the IP block in `ips`.
+
+    SVD files need some basic information that isn't provided within the
+    HJSON descriptions. While they might be available elsewhere in the
+    configuration, it's simplest just to declare them here."""
+
+    # Cross reference module with IP to generate a peripheral
+    def p(m: hjson) -> ET.Element:
+        return peripheral(m, ips[m['type']])
+
+    return create(ET.Element('device', attrib = {
+                'schemaVersion': '1.1',
+                'xmlns:xs': 'http://www.w3.org/2001/XMLSchema-instance',
+                'xs:noNamespaceSchemaLocation': 'CMSIS-SVD.xsd',
+            }),
+            cpu(),
+            create('peripherals', *map(p, top['module'])),
+            vendor          = 'lowRISC OpenTitan',
+            vendorID        = 'lowRISC',
+            name            = top['name'],
+            version         = version,
+            description     = read_relative_path('..', 'README.md'),
+            width           = top['datawidth'],
+            size            = top['datawidth'],
+            addressUnitBits = 8)
+
+def generate(top: hjson, ips: {'name': hjson}, version: str, validate=True) -> ET.Element:
+    """Convert a top-level configuration and IP register definition set
+    into a System View Description ("SVD") file. SVD is an XML format,
+    originally defined by ARM to "formalize the description of the system
+    contained in Arm Cortex-M processor-based microcontrollers".¹ It has
+    since been extended for use on other platforms, and is used in other
+    fields such as Rust embedded programming.²
+
+    SVD is an XML-based format with standardized definitions for memory
+    layouts, register definitions, and similar microcontroller fields.
+
+    The input HJSON must be a top-level configuration (earlgrey) and the
+    full set of IP modeule definitions for that configuration. These two
+    will be parsed and converted into a single XML tree with a top-level
+    "<device>" node. Both the top configuration and IP modules must have
+    been previously validated. (see ??? and `reggen.validate()`)
+
+    If `validate` is true then the resulting SVD file is validated using
+    the `pysvd` module.
+
+    ¹ http://www.keil.com/pack/doc/CMSIS/SVD/html/index.html
+    ² https://github.com/rust-embedded/svd2rust"""
+
+    root = device(top, ips, version)
+
+    # Simply constructing a Device is enough to run the pysvd parser and
+    # validate the structure of the XML tree.
+    if validate:
+        pysvd.element.Device(root)
+
+    # Workaround for limitations of SVD/pysvd. See comment in `cpu()`.
+    root.find('.cpu/name').text = 'RISCV'
+
+    beautify(root)
+    return root
+
+def write_to_file(device: ET.Element, copyright: [str], output):
```

Comment:
This writes to anything with the `write` method, maybe this is just `write_device_stanza`?